web/src/copy_and_paste.js

```diff
@@ -591,15 +614,32 @@ export function paste_handler(event) {
 
         // Only intervene to generate formatted links when dealing
         // with a URL and a URL-safe range selection.
-        if (isUrl(trimmed_paste_text) && is_safe_url_paste_target($textarea)) {
-            event.preventDefault();
-            event.stopPropagation();
-            const url = trimmed_paste_text;
-            compose_ui.format_text($textarea, "linked", url);
-            return;
+        if (isUrl(trimmed_paste_text)) {
+            if (is_safe_url_paste_target($textarea)) {
+                event.preventDefault();
+                event.stopPropagation();
+                const url = trimmed_paste_text;
+                compose_ui.format_text($textarea, "linked", url);
+                return;
+            }
+            if (!compose_ui.cursor_inside_code_block($textarea) && !compose_ui.shift_pressed) {
+                event.preventDefault();
+                event.stopPropagation();
+                // Try to transform the url text to #**stream>topic** syntax
+                // if it is a valid url.
+                const syntax_text = try_stream_topic_syntax_text(trimmed_paste_text, true);
+                if (syntax_text) {
+                    // It is a nice touch to get back the actual pasted url on undoing.
+                    add_text_and_select(trimmed_paste_text, $textarea);
+                }
```

Comment:
I think it'd be cleaner to insert the syntax text and return inside this block itself